reactor-core/src/main/java/reactor/core/publisher/FluxPublish.java

```diff
@@ -392,16 +461,26 @@ PubSubInner<T>[] terminate() {
 		}
 
 		boolean tryConnect() {
-			return connected == 0 && CONNECTED.compareAndSet(this, 0, 1);
+			long previousState = markConnected(this);
+
+			return !isConnected(previousState);
 		}
 
-		final void drain() {
-			if (WIP.getAndIncrement(this) != 0) {
+		final void drainFromInner() {
+			long previousState = addWorkIfSubscribed(this);
+
+			if (!isSubscriptionSet(previousState)) {
 				return;
 			}
 
-			int missed = 1;
+			if (hasWorkInProgress(previousState)) {
+				return;
+			}
 
+			drain(previousState, previousState + 1);
+		}
+
+		final void drain(long previousState, long expectedState) {
```

Comment:
What is the purpose of previousState? It is not used.

Reply:
Fixed. No need for now